### DISCUSSION TITLE
Why are `AsyncThrowingStream` helpers constrained to `Failure == Swift.Error`?

### DISCUSSION THREAD

**Discussion by p4checo** [pointfreeco/swift-dependencies#38] [Q&A] [answered]
hi @stephencelis and @mbrandonw 👋🏼 

Could you please share the reasoning for the `AsyncThrowingStream` helpers to be constrained to `Failure == Swift.Error`?

Is there any limitation around using custom error types? I have a faint idea of having read about it _somewhere_, but I've searched the documentation, GH repos and git history and found no reference of that. Apologies if I missed it.

If there's a limitation then I guess it would be good if it were documented. Otherwise, does it make sense for this restriction be lifted?

Thanks! 🙇🏼

**Comment by stephencelis** [comment, marked as answer]
@p4checo We're following Apple's lead here: https://developer.apple.com/documentation/swift/asyncthrowingstream/init(_:bufferingpolicy:_:)

They're generic over `Failure` to prepare for a future where typed throws is supported, but till then it can only be equal to the `Error` existential.

**Reply by p4checo** [reply]
Thanks @stephencelis! 

Everything makes much more sense now... 😅 I completely missed that restriction on the main `init`.

Typed throws will IMO fill in a big gap in the language and remove quite a bit of friction. Can't wait for that!
